### PR TITLE
Override autoscale

### LINKED ISSUE #382: Create Grid instance for every pane
For now we use the only one Grid instance per the whole chart, but create pane view for every pane, what led by [using WeakMap to avoid memory leaks](https://github.com/tradingview/lightweight-charts/blob/5e328f4982675c78d8a7599a240a7d108cc84034/src/model/grid.ts#L26-L31). If we'll create Grid instance per pane, we don't need to use WeakMap, so -1 required polyfill 🙂.

### LINKED ISSUE #392: Add ability to modify price scale's price range
**Is your feature request related to a problem? Please describe.**

Yes, I want to be able to set the price scale's min and max value displayed.

**Describe the solution you'd like**

chart.setOption({ priceScale: { min: <value>, max: <value> } })

### PR DESCRIPTION
**Pull request by eugene-korobko**
**Type of PR:** enhancement

**PR checklist:**

- [x] Addresses an existing issue: fixes #392
- [x] Includes tests
- [x] Documentation update

**Overview of change:**
Now there is an option to override series autoscale value.

### CODE REVIEW

File: docs/series-basics.md [hunk 1]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+            priceRange: {
+                minValue: 0,
+                maxValue: 100,
+            },
+        };
+    },
+});
+```
+
+So, you could just add a function that returns object with `priceRange` field to make chart thinking the series always has values from 0 to 100. These values aer in prices.
```
Comment (timocov) [inline docs/series-basics.md:139]:
```suggestion
So, you could just add a function that returns object with `priceRange` field to make chart thinking the series always has values from 0 to 100. These values are in prices.
```

File: docs/series-basics.md [hunk 2]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+            margins: {
+                above: 10,
+                below: 10,
+            },
+        };
+    },
+});
+```
+
+Actually, `autoscaleProvider` function accepts a parameter with original implementation, so you can call default and ajust the result:
```
Comment (timocov) [inline docs/series-basics.md:160]:
```suggestion
Actually, `autoscaleProvider` function accepts a parameter with original implementation, so you can call default and adjust the result:
```

File: docs/series-basics.md [hunk 3]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+        if (res.priceRange != null) {
+            res.priceRange.minValue -= 10;
+            res.priceRange.maxValue += 10;
+        }
+        return res;
+    },
+});
+```
+
+Please pay your attention that both `priceRange` and `margins` could be null in the default result.
```
Comment (timocov) [inline docs/series-basics.md:175]:
```suggestion
Please pay your attention that both `priceRange` and `margins` could be `null` in the default result.
```

File: docs/series-basics.md [hunk 4]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+});
+```
+
+Actually, `autoscaleProvider` function accepts a parameter with original implementation, so you can call default and ajust the result:
+
+```javascript
+var firstSeries = chart.addLineSeries({
+    autoscaleProvider: (original) => {
+        var res = original();
+        if (res.priceRange != null) {
```
Comment (timocov) [inline docs/series-basics.md:166]:
```suggestion
        if (res.priceRange !== null) {
```

File: src/model/series-options.ts [hunk 5]
```diff
@@ -163,6 +164,20 @@ export interface NonOverlaySeriesSpecificOptions {
...
+}
+
+export interface AutoscaleInfo {
+	priceRange: PriceRange | null;
+	margins: AutoScaleMargins | null;
+}
+
+type AutoscaleProviderImplementation = () => AutoscaleInfo | null;
+
+type AutoscaleProvider = (baseImplementation: AutoscaleProviderImplementation) => AutoscaleInfo | null;
```
Comment (timocov) [inline src/model/series-options.ts:179]: AutoscaleInfoProvider?

File: src/model/series-options.ts [hunk 6]
```diff
@@ -163,6 +164,20 @@ export interface NonOverlaySeriesSpecificOptions {
...
+	minValue: number;
+	maxValue: number;
+}
+
+export interface AutoscaleInfo {
+	priceRange: PriceRange | null;
+	margins: AutoScaleMargins | null;
+}
+
+type AutoscaleProviderImplementation = () => AutoscaleInfo | null;
```
Comment (timocov) [inline src/model/series-options.ts:177]: Maybe reduce that type and just inline it in AutoscaleProvider? It's name is a bit strange :)

File: src/model/series.ts [hunk 7]
```diff
@@ -395,33 +395,16 @@ export class Series<T extends SeriesType = SeriesType> extends PriceDataSource i
...
-		let barsMinMax: MinMax | null;
-		if (priceSource !== null) {
-			barsMinMax = this.data().bars().minMaxOnRangeCached(startTimePoint, endTimePoint, [{ name: priceSource, offset: 0 }]);
-		} else {
-			barsMinMax = this.data().bars().minMaxOnRangeCached(startTimePoint, endTimePoint, [{ name: 'low', offset: 0 }, { name: 'high', offset: 0 }]);
-		}
-
-		let range = barsMinMax !== null ? new PriceRange(barsMinMax.min, barsMinMax.max) : null;
+	public autoscaleInfo(startTimePoint: TimePointIndex, endTimePoint: TimePointIndex): AutoscaleInfoImpl | null {
+		if (this._options.autoscaleProvider) {
```
Comment (timocov) [inline src/model/series.ts:399]:
```suggestion
		if (this._options.autoscaleProvider !== undefined) {
```

File: src/model/series-options.ts (hunk 5 above)
Comment (eugene-korobko) [reply src/model/series-options.ts]: ok

File: src/model/series-options.ts (hunk 6 above)
Comment (eugene-korobko) [reply src/model/series-options.ts]: done

File: docs/series-basics.md [hunk 8]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
 ```
 
+### Overriding autoscale
+
+By default, chart scales data automatically, based on visible range of data. However for some reasons one could need overriding this behavior.
+There is an options called `autoscaleProvider` that allows overriding visible price range for series
+
+```javascript
+var firstSeries = chart.addLineSeries({
+    autoscaleProvider: () => {
```
Comment (timocov) [inline docs/series-basics.md:128]:
```suggestion
    autoscaleInfoProvider: () => {
```

File: docs/series-basics.md [hunk 9]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+});
+```
+
+So, you could just add a function that returns object with `priceRange` field to make chart thinking the series always has values from 0 to 100. These values are in prices.
+
+You could also provide additional margins in pixels. Please be careful and never mix prices and pixels.
+
+```javascript
+var firstSeries = chart.addLineSeries({
+    autoscaleProvider: () => {
```
Comment (timocov) [inline docs/series-basics.md:145]:
```suggestion
    autoscaleInfoProvider: () => {
```

File: docs/series-basics.md [hunk 10]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+            margins: {
+                above: 10,
+                below: 10,
+            },
+        };
+    },
+});
+```
+
+Actually, `autoscaleProvider` function accepts a parameter with original implementation, so you can call default and adjust the result:
```
Comment (timocov) [inline docs/series-basics.md:160]:
```suggestion
Actually, `autoscaleInfoProvider` function accepts a parameter with original implementation, so you can call default and adjust the result:
```

File: docs/series-basics.md [hunk 11]
```diff
@@ -118,6 +118,62 @@ chart.applyOptions({
...
+
+```javascript
+var firstSeries = chart.addLineSeries({
+    autoscaleProvider: (original) => {
```
Comment (timocov) [inline docs/series-basics.md:164]:
```suggestion
    autoscaleInfoProvider: (original) => {
```

**Pull request closed without merging**